rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

```diff
@@ -0,0 +1,95 @@
+| Bincode            |   5882 |       8772 |     9524 |     4016 |          163 |         129 |
+| Protobuf           |    859 |       1140 |     1163 |      480 |          154 |         141 |
+| Prost              |   1225 |       1866 |     1984 |      962 |          154 |         142 |
+
+Performance is a single aspect of a nuanced problem for us; for more perf benchmarks I'd suggest checking out [rust-serialization-benchmarks](https://github.com/erickt/rust-serialization-benchmarks).  Some of these formats (JSON for example) still leaves us with a marginal concern around both dropped metadata as well as incidental exposure of a private data format to our users. The goal here is to hopefully solve for all of these things with whatever path is chosen.
+
+While it is tempting to look at perf benchmarks and to just pick the best performing example, the practical requirements around the problem for us might give us some limitations. As was [expressed by @lukesteensen](https://discord.com/channels/742820443487993987/746070604192415834/796425792875397120) in the #development channel in discord we have three specific concerns around the choice:
+
+-  Our internal data model itself (inclusive of our desire to not incidentally create _another_ observability data format)
+-  How we serialize that data for internal use
+-  The api for accessing that data model (e.g. lookups)
+
+These three concerns don't fully enumerate the problem but I think they're a really excellent distillation of what we need to be considering. So, to reiterate - our ideal solution solves for all three but that might just not be possible. While I initially investigated this problem with really only #2 in mind, discussions with folks on the team led me to feeling like a real firecracker of a solution might be possible for both #1 and #2 out box. With the current state of lookups today, I'm less inclined to try to solve #3 immediately and I'm unconvinced that something might provide us a solution to it directly out of box. (Though, with that said, [flatbuffers](https://google.github.io/flatbuffers/) might be an interesting piece of software to take a look at in the future since it provides users the ability to access serialized data without unpacking it despite it not ultimately being my suggested approach.)
+
+It might not be totally clear from the serialization benchmark examples but while protobufs are inherently slow in some respects, they're _still_ more performant than raw JSON implementations and even MssagePack (which I would be willing to bet remains true for CBOR). Any schema based data format is likely going to have better perf than something schemaless. The schema itself is obviously part of the tooling overhead here but I'd argue that if we cared _more_ about tooling overhead than we did about perfomance and sustainability we'd probably be writing this project in something other than Rust.
```

Comment:
A big :+1: on this. It's also worth noting that the tooling overhead for protobuf is largely a cost we've already paid.

Reply:
I'd like to point out that the schema that we have currently in nowhere near in complexity to the use cases that protobuf was designed for - in a sense that we could possible hand-write the whole of the serialization logic with some optimal, specially tailored thing - if only we could (knew how to) do it better than protobuf and other autogenerated solutions.
The problem is it doesn't seem to be worth doing. Although... It might worth a try. At least we'll have a greater degree of control over performance.

The downside of the hard-written solution (which is really the upside of the "tooling") is that instead of tooling overhead we'll have some hand-written code - which has a maintenance cost, potentially higher than the tooling overhead, although it feels less "hacky".

Reply:
I'll add this to the potential alternatives.